File: Codes/Programmers/조이스틱.py
import sys

sys.setrecursionlimit(100000)


def solution(name):
    # A로 바꾸는데 드는 비용
    def count2A(char):
        return min(ord(char) - 65, 91 - ord(char))

    # i: 현재 인덱스, cnt: 이동 횟수, A: A의 개수
    def dfs(i, cnt, A):
        nonlocal min_value
        if A == n:
            min_value = min(min_value, cnt)
            return
        # 가지치기
        if cnt >= min_value:
            return

        # 오른쪽으로 A 아닌 곳
        j = i
        while True:
            j += 1
            if not visited[j % n]:
                visited[j % n] = True
                dfs(j % n, cnt + j - i + count2A(name[j % n]), A + 1)
                visited[j % n] = False
                break

        # 왼쪽으로 A 아닌 곳
        j = i
        while True:
            j -= 1
            if not visited[j % n]:
                visited[j % n] = True
                dfs(j % n, cnt + i - j + count2A(name[j % n]), A + 1)
                visited[j % n] = False
                break

    n = len(name)
    visited = [False] * n  # A인가?
    a = 0
    for idx, char in enumerate(name):
        if char == "A":
            visited[idx] = True
            a += 1

    min_value = 1e10

    if name[0] == "A":
        dfs(0, 0, a)
    else:
        visited[0] = True
        dfs(0, count2A(name[0]), a + 1)

    return min_value

File: Codes/Programmers/조이스틱.js
function solution(name) {
  // A로 만드는 비용
  const toMakeA = (char) =>
    Math.min(char.charCodeAt() - 65, 91 - char.charCodeAt());

  const n = name.length;
  const visited = Array(n).fill(false); // A 인가요?
  let aCount = 0;
  let minValue = Infinity;

  [...name].forEach((char, index) => {
    if (char === "A") {
      visited[index] = true;
      aCount++;
    }
  });

  function dfs(i, count, A) {
    if (A === n) {
      minValue = Math.min(minValue, count);
      return;
    }

    // 가지치기
    if (count >= minValue) return;

    // 오른쪽으로 가서 A 아닌 곳 찾기
    let j = i;
    while (true) {
      j++;
      const nj = (j + n) % n;
      if (!visited[nj]) {
        visited[nj] = true;
        dfs(nj, count + j - i + toMakeA(name[nj]), A + 1);
        visited[nj] = false;
        break;
      }
    }

    // 왼쪽으로 가서 A 아닌 곳 찾기
    j = i;
    while (true) {
      j--;
      const nj = (j + n) % n;
      if (!visited[nj]) {
        visited[nj] = true;
        dfs(nj, count + i - j + toMakeA(name[nj]), A + 1);
        visited[nj] = false;
        break;
      }
    }
  }

  if (name[0] === "A") {
    dfs(0, 0, aCount);
  } else {
    visited[0] = true;
    dfs(0, toMakeA(name[0]), aCount + 1);
  }

  return minValue;
}
